python-lib/dku_email_client.py

```diff
@@ -169,10 +172,16 @@ def send_single_email(self, sender, recipients, cc_recipients, bcc_recipients, e
         for mime_app in attachment_mimes:
             msg.attach(mime_app)
 
-        self.smtp.sendmail(sender, all_recipients, msg.as_string())
+        try:
+            self.smtp.sendmail(sender, all_recipients, msg.as_string())
+        finally:
+            # Explicitly reset the session or "Error: nested MAIL command" error is possible
+            self.smtp.rset()
 
     def send_email(self, sender, recipients, cc_recipients, bcc_recipients, email_subject, email_body, attachment_files):
         attachment_mimes = self.attachments_to_mime(attachment_files)
+        if not recipients:
+            raise Exception("No recipients provided to send to")
```

Comment:
Explaining patch changes: again we get a much improved error message if we throw here - the error messages had become a bit cryptic due to the other changes